File: 0028-find-the-index-of-the-first-occurrence-in-a-string/0028-find-the-index-of-the-first-occurrence-in-a-string.cpp
class Solution {
public:
    int strStr(string haystack, string needle) {
        
        const int pow = 31; 
        
        const int mod = 1e9 + 9;
        
        int S = needle.size(), T = haystack.size();

        vector<long long> p_pow(max(S, T)); 
        
        p_pow[0] = 1; 
        
        for (int i = 1; i < (int)p_pow.size(); i++) 
            p_pow[i] = (p_pow[i-1] * pow) % mod;

        vector<long long> h(T + 1, 0); 
        
        
        for (int i = 0; i < T; i++)
            
            h[i+1] = (h[i] + (haystack[i] - 'a' + 1) * p_pow[i]) % mod; 
        
        long long h_n = 0; 
        for (int i = 0; i < S; i++) 
            h_n = (h_n + (needle[i] - 'a' + 1) * p_pow[i]) % mod; 

 
        
        int ans = -1;
        
        for (int i = 0; i + S - 1 < T; i++) {
            
            long long cur_h = (h[i+S] + mod - h[i]) % mod;
            
            if (cur_h == h_n * p_pow[i] % mod){
                ans = i;
                break;
            }
        }
        return ans;
    }
};

File: 0028-find-the-index-of-the-first-occurrence-in-a-string/0028-find-the-index-of-the-first-occurrence-in-a-string.py
class Solution:
    
    def strStr(self, haystack: str, needle: str) -> int:
        
        sz_hay = len(haystack)
        
        sz_nee = len(needle)

        p = 26
        
        m = 10 ** 9 + 9
        
        ascii_a = ord('a')
        
        pow = [0] * max(sz_hay, sz_nee)
        
        
        
        sz_pow = len(pow)

        pow[0] = 1
        
        for i in range(1,sz_pow):
            
            pow[i] = (pow[i-1] * p) % m
            
        # print(pow)
        
        hash = [0] * (sz_hay+1)

        for i in range(sz_hay):
            
            hash[i+1] = (hash[i] + (ord(haystack[i]) - ascii_a + 1) * pow[i]) % m
            
        
        hash_needle = 0

        for i in range(sz_nee):
            
            hash_needle = (hash_needle + (ord(needle[i]) - ascii_a + 1) * pow[i]) % m
            # print(hash_needle)
        
        stat = -1

        for i in range(sz_hay - sz_nee + 1):
            
            cur_hash = (hash[i+sz_nee] + m - hash[i]) % m
            
            if (hash_needle * pow[i]) % m == cur_hash:
                
                stat = i
                
                break

        return stat